### PR TITLE
Books index show tests

### PR DESCRIPTION
**Pull request by mpuricawolox**
## Summary

Adding tests for books controller

## Trello Card

[https://trello.com/c/dZIplZhg/25-crear-tests-para-rutas-index-y-show-de-book](https://trello.com/c/dZIplZhg/25-crear-tests-para-rutas-index-y-show-de-book)

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -10,4 +10,10 @@ def configure_permitted_parameters
   end
 
   protect_from_forgery with: :null_session
+
+  rescue_from ActiveRecord::RecordNotFound, with: :handle_record_not_found
+
+  def handle_record_not_found
+    render(json: { error: 'Nothing found' }, status: 404)
```
Comment (leamotta) [inline app/controllers/application_controller.rb]: we can use status code symbols here to make it more readable. For 404 it's `:not_found`. You can find a list here http://billpatrianakos.me/blog/2013/10/13/list-of-rails-status-code-symbols/

File: spec/controllers/api/v1/books_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    context 'when filtering books by genre' do
+      let!(:drama) { create_list(:drama, 3) }
+      let(:books) do
+        arr = create_list(:book, 3)
+        arr.push(*drama)
+      end
```
Comment (leamotta) [inline spec/controllers/api/v1/books_controller_spec.rb]: there's an easier way for this. The `create` and `create_list` method support fixed parameters, so you could do
`let(:books) { create_list(:book, 3, genre: 'drama') }` To create 3 different drama books

Comment (leamotta) [reply spec/controllers/api/v1/books_controller_spec.rb]: Also, it's better to also create some random books with other random genres to test that the filter actually works. If you only create the drama books the filter could be broken and this test would pass anyway

File: spec/controllers/api/v1/books_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    end
+
+    context 'when fetching a book with invalid id' do
+      let(:book) { create(:book) }
+
+      before do
+        get :show, params: { id: 2 }
+      end
+
+      it 'responds with the book json' do
```
Comment (leamotta) [inline spec/controllers/api/v1/books_controller_spec.rb]: `responds with error` here :D

File: spec/rails_helper.rb [hunk 4]
```diff
@@ -29,6 +29,15 @@
   puts e.to_s.strip
   exit 1
 end
+
+RSpec.shared_context 'Authenticated User' do
+  let(:user) { create(:user) }
```
Comment (leamotta) [inline spec/rails_helper.rb]: let's put this in a separate file, so it's clearly visible and refactorable.
For example, in spec/support/shared_contexts/authenticated_user.rb

File: spec/factories/book_factory.rb [hunk 5]
```diff
@@ -6,5 +6,8 @@
     title { Faker::Lorem.words }
     publisher { Faker::Name.name }
     year { Faker::Number.number(4) }
+    factory :drama do
+      genre { 'drama' }
```
Comment (leamotta) [inline spec/factories/book_factory.rb]: this is unnecessary, check my comment in the tests

**Review by leamotta — CHANGES_REQUESTED** [review]
A few handy things to make it easier in the future

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (mpuricawolox) [reply app/controllers/application_controller.rb]: Greate, I thought they existed only in specs

**Review by leamotta — APPROVED** [review]
🎖

**Pull request merged**